src/file-chooser.c

```diff
@@ -176,15 +229,27 @@ handle_open_file (XdpFileChooser *object,
                                                   arg_parent_window,
                                                   arg_title,
                                                   g_variant_builder_end (&options),
-                                                  &impl_handle,
+                                                  request->id,
                                                   NULL, &error))
     {
       g_dbus_method_invocation_return_gerror (invocation, error);
       return TRUE;
     }
 
-  g_object_set_data_full (G_OBJECT (request), "impl-handle", g_strdup (impl_handle), g_free);
-  register_handle (impl_handle, request);
+  impl_request = xdp_impl_request_proxy_new_sync (g_dbus_proxy_get_connection (G_DBUS_PROXY (impl)),
+                                                  G_DBUS_PROXY_FLAGS_NONE,
+                                                  g_dbus_proxy_get_name (G_DBUS_PROXY (impl)),
+                                                  request->id,
+                                                  NULL, &error);
+  if (!impl_request)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  g_signal_connect (impl_request, "response", (GCallback)handle_open_file_response, request);
+
+  g_object_set_data_full (G_OBJECT (request), "impl-request", g_object_ref (impl_request), g_object_unref);
```

Comment:
We could even add a `GDBusProxy *impl_request` to `Request` to avoid the set/get_data() calls.